### PR TITLE
[Datasets] Provide more efficient + intuitive block clearing semantics for different execution modes

### LINKED ISSUE #18791: Optimize memory usage of datasets
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

Currently, Datasets are considered immutable. Each operation (e.g., `.map`) creates new data blocks, leaving the original data block references untouched.

However, this can lead to 2x+ memory usage in the cluster, since Ray cannot safely delete the old blocks, and both block references are at least referenced until the operation finishes (e.g., in `ds = ds.map(f)`).

There are a few options to resolve this:

**Option 1:** Have all operations work in-place (mutate the dataset). Add a .clone() method if users want to get the existing behavior.

```
ds = ray.data.range(1000)
ds.map(f)
ds.random_shuffle()  # assignment to ds is optional, the original copy is mutated
```

**Option 2:** Have a lazy mode (ds.lazy()) where operations are queued and executed lazily.

```
ds = ray.data.range(1000)
ds = ds.map(f)
ds = ds.random_shuffle()
ds.take()  # actually executes the lazy operation
```

**Option 3:** Allow users to select whether operations are in-place/move blocks per operation.

```
ds = ray.data.range(1000)
ds.map_(f)
ds.random_shuffle_()  # pytorch-like "inplace" syntax
```

**Option 4:** Rely on lineage reconstruction, and auto delete past blocks by default.

```
ds0 = ray.data.range(1000)
ds = ds0.map(f)  # drops ds0 blocks from memory
ds = ds0.random_shuffle()  # reconstructs original ds before shuffling
```
----

Currently the thought is that Option 4 is the best, but relies on lineage reconstruction to be on by default, as well as proper ray.put() support (cc @stephanie-wang )

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by clarkzinzow**
**TL;DR:** Don't clear for eager, clear all but non-lazy input blocks if lazy, clear everything if pipelining.
 
This PR provides more efficient and intuitive block clearing semantics for eager mode, lazy mode, and pipelining, while still supporting multiple operations applied to the same base dataset, i.e. fan-out. For example, two different map operations are applied to the same base `ds` in this example:

```python
ds = ray.data.range(10).map(lambda x: x+1)
ds1 = ds.map(lambda x: 2*x)
ds2 = ds.map(lambda x: 3*x)
```

If naively clear the blocks when executing the map to produce `ds1`, the map producing `ds2` will fail.

### Desired Semantics

- **Eager mode** - don’t clear input blocks, thereby supporting fan-out from cached data at any point in the stage chain without triggering unexpected recomputation.
- **Lazy mode** - if lazy datasource, clear the input blocks for every stage, relying on recomputing via stage lineage if fan-out occurs; if non-lazy datasource, do not clear source blocks for execution plan when executing first stage, but do clear input blocks for every subsequent stage.
- **Pipelines** - Same as lazy mode, although the only fan-out that can occur is from the pipeline source blocks when repeating a dataset/pipeline, so unintended intermediate recomputation will never happen.

## Related issue number

Closes #18791 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/impl/plan.py [hunk 1]
```diff
@@ -224,23 +224,33 @@ def meta_count(self) -> Optional[int]:
             return None
 
     def execute(
-        self, clear_input_blocks: bool = True, force_read: bool = False
+        self,
+        clear_input_blocks: bool = True,
+        destroy_unrecoverable_input_blocks: bool = False,
```
Comment (ericl) [inline python/ray/data/impl/plan.py:229]: I like the name of this argument!

**Review by ericl — COMMENTED** [review]
LGTM at a high level. @jianoaix can you review?

**Comment by jjyao** [comment]
What's `fan-out` here?

**Comment by clarkzinzow** [comment]
@jjyao By fan-out, we mean more than one dataset operation applied to a single base dataset, e.g. two different map operations applied to the same base `ds` in this example:
```python
ds = ray.data.range(10).map(lambda x: x+1)
ds1 = ds.map(lambda x: 2*x)
ds2 = ds.map(lambda x: 3*x)
```

I'll add this to the PR description!

**Comment by jianoaix** [comment]
Actually, with fan-out, I think there is an issue with execution plan, which will be a DAG, rather than a chain.

For example, continue the code snippet, we can have a fan-in: ds3 = ds1.union(ds2). Now the lineage of ds3 is a DAG. And user can continue such pattern to construct complex DAG.

**Comment by clarkzinzow** [comment]
@jianoaix Agreed that an explicit DAG representation would be more ideal (that's the future plan), but what do you see as the concrete current issue with fan-out here? Fan-out is supported in eager mode via reusing the base dataset's snapshot blocks, and is supported in lazy mode via recomputation from source, so fan-out should be covered. I.e., the stage DAG for fan-out of lazy datasets is currently supported by independent chains for each dataset sink, with the chain going back to the shared source dataset.

>For example, continue the code snippet, we can have a fan-in: ds3 = ds1.union(ds2). Now the lineage of ds3 is a DAG. And user can continue such pattern to construct complex DAG.

Right now, lazy fan-in isn't supported: all fan-in operations trigger computation for lazy datasets. One thing that's currently untested is lazy fan-out after a union. Upon a quick eye-test, it looks like it might not work if one or more of the unioned datasets are created from non-lazy datasources, since recomputing from the base won't work on the [dummy read tasks](https://github.com/ray-project/ray/blob/a7a6465936950bc8155b6fb658cd362571f137f4/python/ray/data/dataset.py#L1008).

**Comment by jianoaix** [comment]
With the lineage/execution plan being used for both eager and lazy dataset (e.g. the other PR claimed the eager dataset can also serialize the lineage), isn't this already a problem for both cases using linear data structure for DAG?

**Comment by clarkzinzow** [comment]
Synced offline, opened [this PR](https://github.com/ray-project/ray/pull/24190) to make it explicit that lineage-based serialization of unions/zips are not supported.

File: python/ray/data/impl/pipeline_executor.py [hunk 2]
```diff
@@ -19,7 +19,9 @@ def pipeline_stage(fn: Callable[[], Dataset[T]]) -> Dataset[T]:
     # Force eager evaluation of all blocks in the pipeline stage. This
     # prevents resource deadlocks due to overlapping stage execution (e.g.,
     # task -> actor stage).
-    return fn().fully_executed()
+    # We allow non-lazy (unrecoverable) input blocks to be destroyed here since no
+    # fan-out is allowed in a pipeline.
+    return fn().fully_executed(preserve_input_blocks=True)
```
Comment (jianoaix) [inline python/ray/data/impl/pipeline_executor.py:24]: It looks this should be False?

File: python/ray/data/impl/plan.py [hunk 3]
```diff
@@ -224,23 +224,33 @@ def meta_count(self) -> Optional[int]:
...
+        force_read: bool = False,
     ) -> BlockList:
         """Execute this plan.
 
         Args:
-            clear_input_blocks: Whether to assume ownership of the input blocks,
```
Comment (jianoaix) [inline python/ray/data/impl/plan.py:232]: This parameter is still in the signature, it should be retained.

File: python/ray/data/impl/plan.py [hunk 4]
```diff
@@ -293,20 +335,32 @@ def _optimize(self) -> Tuple[BlockList, DatasetStats, List[Stage]]:
...
-        """Get the source blocks (and corresponding stats) for plan execution.
+    def _get_source_blocks_and_stages(
+        self,
+    ) -> Tuple[BlockList, DatasetStats, List[Stage]]:
+        """Get the source blocks, corresponding stats, and the stages for plan
+        execution.
 
-        If a computed snapshot exists, return the snapshot blocks and stats; otherwise,
-        return the input blocks and stats that the plan was created with.
+        If a computed snapshot exists and has not been cleared, return the snapshot
```
Comment (jianoaix) [inline python/ray/data/impl/plan.py:344]: In retrospective, I think this essentially is (and could probably be called) "watermark" (of plan execution) - and then we can unify the two cases, be it input blocks (watermark at 0) or snapshot blocks (watermark at position N >= 1). (No need to change in this PR)

File: python/ray/data/tests/test_optimize.py [hunk 5]
```diff
@@ -83,6 +94,67 @@ def test_memory_release_lazy_shuffle(shutdown_only):
     raise error
 
 
+def test_lazy_fanout(shutdown_only, local_path):
```
Comment (jianoaix) [inline python/ray/data/tests/test_optimize.py:97]: Nice test!

File: python/ray/data/tests/test_optimize.py [hunk 6]
```diff
@@ -83,6 +94,67 @@ def test_memory_release_lazy_shuffle(shutdown_only):
...
+        {"one": 3, "two": "a"},
+        {"one": 4, "two": "b"},
+        {"one": 5, "two": "c"},
+    ]
+    assert ds3.fully_executed().take() == [
+        {"one": 3, "two": "a"},
+        {"one": 4, "two": "b"},
+        {"one": 5, "two": "c"},
+    ]
+    # Test that data is read twice (+ 1 extra for ramp-up read before converting to a
```
Comment (jianoaix) [inline python/ray/data/tests/test_optimize.py:131]: hm, can it be fully lazy and avoid this extra?

File: python/ray/data/impl/plan.py (hunk 4 above)
Comment (clarkzinzow) [reply python/ray/data/impl/plan.py:344]: Hmm I'm used to "watermark" being used for streaming systems in terms of handling "late" data, can you point me to where it's used in this context? I also feel like snapshot is a pretty accurate name, but happy to discuss alternatives.

File: python/ray/data/impl/pipeline_executor.py (hunk 2 above)
Comment (clarkzinzow) [reply python/ray/data/impl/pipeline_executor.py:24]: Hmm great catch, this should be covered by tests as well. 🤔 I'll fix this and make sure it's covered by tests.

File: python/ray/data/tests/test_optimize.py [hunk 7]
```diff
@@ -29,23 +59,92 @@ def test_memory_sanity(shutdown_only):
...
+    for block in pipe.iter_batches(batch_size=None):
+        for arr in block:
+            np.testing.assert_equal(
+                arr,
+                np.ones(100 * 1024 * 1024, dtype=np.uint8) + num_rounds,
+            )
     meminfo = memory_summary(info["address"], stats_only=True)
-    # TODO(ekl) we can add this back once we clear inputs on eager exec as well.
-    # assert "Spilled" not in meminfo, meminfo
+    assert "Spilled" not in meminfo, meminfo
```
Comment (clarkzinzow) [inline python/ray/data/tests/test_optimize.py:141]: This test might prove to be flaky due to the memory reclaiming issues within Ray, so we may need to put this in a retry loop similar to `test_memory_release_lazy_shuffle`.

File: python/ray/data/dataset.py [hunk 8]
```diff
@@ -2861,19 +2860,27 @@ def __iter__(self):
...
 
-    def fully_executed(self) -> "Dataset[T]":
+    def fully_executed(self, preserve_original: bool = True) -> "Dataset[T]":
         """Force full evaluation of the blocks of this dataset.
 
         This can be used to read all blocks into memory. By default, Datasets
         doesn't read blocks from the datasource until the first transform.
 
+        Args:
+            preserve_original: Whether the original unexecuted dataset should be
```
Comment (jianoaix) [inline python/ray/data/dataset.py:2870]: What is "original dataset" for a dataset? (I think users may have same question if they read this API)

File: python/ray/data/impl/plan.py (hunk 4 above)
Comment (jianoaix) [reply python/ray/data/impl/plan.py:337]: I think 'snapshot' as something that won't change whereas here we are tracking progress of a moving cursor. But I'm closing this thread since it's not public API (and not ask for a change in this PR), we can chat separately.

File: python/ray/data/impl/plan.py [hunk 9]
```diff
@@ -275,13 +285,45 @@ def stats(self) -> DatasetStats:
         self.execute()
         return self._snapshot_stats
 
+    def _should_clear_input_blocks(
+        self,
+        blocks: BlockList,
+        stage_idx: int,
+        clear_input_blocks: bool,
```
[the remark this answers is not in the document]
Comment (jianoaix) [reply python/ray/data/impl/plan.py]: Does this make sense (since I didn't see response or change)? Basically my thinking is a function like "bool do_something(bool dont_do_if_true, others)" should remove "dont_do_if_true" out of the parameters.

File: python/ray/data/tests/test_optimize.py [hunk 10]
```diff
@@ -29,23 +59,92 @@ def test_memory_sanity(shutdown_only):
...
-    meminfo = memory_summary(info.address_info["address"], stats_only=True)
-    assert "Spilled" not in meminfo, meminfo
-
-    # Round 2.
-    ds = ds.map(lambda x: np.ones(100 * 1024 * 1024, dtype=np.uint8))
+    def gen(x):
+        import time
+
+        # TODO(Clark): Remove this sleep once we have fixed memory pressure handling.
+        time.sleep(2)
```
Comment (jianoaix) [inline python/ray/data/tests/test_optimize.py:115]: File a bug to track this issue?

File: python/ray/data/tests/test_optimize.py (hunk 7 above)
Comment (jianoaix) [reply python/ray/data/tests/test_optimize.py:141]: Ack. Even 2s not enough to escape the reclaiming uncertainty?

**Review by jianoaix — COMMENTED** [review]
LGTM, just a few minor comments

**Review by jianoaix — APPROVED** [review]

File: python/ray/data/impl/plan.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/impl/plan.py]: Yep I gave a 👍 and made the change!

File: python/ray/data/dataset.py (hunk 8 above)
Comment (clarkzinzow) [reply python/ray/data/dataset.py:2870]: Ah I thought that when calling `d2 = ds.fully_executed(preserve_original=False)`, `original` referring to `ds` would be obvious. 🤔

File: python/ray/data/tests/test_optimize.py (hunk 10 above)
Comment (clarkzinzow) [reply python/ray/data/tests/test_optimize.py:115]: Yep can do!

File: python/ray/data/tests/test_optimize.py (hunk 7 above)
Comment (clarkzinzow) [reply python/ray/data/tests/test_optimize.py:141]: 1s was a bit flaky, didn't see 2s fail while testing locally, but who knows in CI. For context, we use an "OOM grace period" of [2 seconds](https://github.com/ray-project/ray/blob/9f88031d4fff7547c884ec304766acd41d73989f/src/ray/common/ray_config_def.h#L428-L430) in core Ray, where we let reference counting + garbage collection and object spilling free up space, so I'd expect this to be ~enough.

**Review by ericl — APPROVED** [review]

File: python/ray/data/dataset.py (hunk 8 above)
Comment (jianoaix) [reply python/ray/data/dataset.py:2870]: I see. I was having a chain of operations in mind, in that case it's unclear which one of them should be thought as "original" (maybe the first one? maybe each intermediate ones? or others?).
